File: unity/simple-stack-vm-unity/Assets/Scripts/SimpleStackVM/Values/BoolValue.cs
#nullable enable

namespace SimpleStackVM
{
    public struct BoolValue : IValue
    {
        #region Fields
        public static BoolValue True = new BoolValue(true);
        public static BoolValue False = new BoolValue(false);

        public readonly bool Value;

        public string TypeName => "bool";
        #endregion

        #region Constructor
        public BoolValue(bool value)
        {
            this.Value = value;
        }
        #endregion

        #region Methods
        public override string ToString() => this.Value ? "true" : "false";

        public int CompareTo(IValue? other)
        {
            if (other == null) return 1;
            if (other is BoolValue otherBoolValue)
            {
                return this.Value.CompareTo(otherBoolValue.Value);
            }

            return 1;
        }
        #endregion
    }
}

File: ts/src/values/boolValue.ts
import { CompareResult, IValue } from "./ivalues";

export default class BoolValue implements IValue
{
    public static readonly True: BoolValue = new BoolValue(true);
    public static readonly False: BoolValue = new BoolValue(false);

    public readonly value: boolean;

    constructor (value: boolean)
    {
        this.value = value;
    }

    public typename() { return 'boolean'; }
    public toString() { return this.value ? 'true' : 'false'; }

    public compareTo(other: IValue): CompareResult
    {
        if (!(other instanceof BoolValue))
        {
            return 1;
        }

        return boolCompareTo(this.value, other.value);
    }
}

export function boolCompareTo(left: boolean, right: boolean): CompareResult
{
    if (left === right)
    {
        return 0;
    }
    return left ? -1 : 1;
}

export function isBoolValue(input: IValue | undefined): input is BoolValue
{
    return input instanceof BoolValue;
}